Library/Homebrew/formula.rb

```diff
@@ -2214,161 +2214,59 @@ def merge_spec_dependables(dependables)
 
   # @private
   def to_hash
-    # Create a hash of spec names (stable/head) to the list of dependencies under each
-    dependencies = self.class.spec_syms.to_h do |sym|
-      [sym, send(sym)&.declared_deps]
-    end
-    dependencies.transform_values! { |deps| deps&.reject(&:implicit?) } # Remove all implicit deps from all lists
-    requirements = self.class.spec_syms.to_h do |sym|
-      [sym, send(sym)&.requirements]
-    end
-
     hsh = {
-      "name"                     => name,
-      "full_name"                => full_name,
-      "tap"                      => tap&.name,
-      "oldname"                  => oldnames.first, # deprecated
-      "oldnames"                 => oldnames,
-      "aliases"                  => aliases.sort,
-      "versioned_formulae"       => versioned_formulae.map(&:name),
-      "desc"                     => desc,
-      "license"                  => SPDX.license_expression_to_string(license),
-      "homepage"                 => homepage,
-      "versions"                 => {
+      "name"                   => name,
+      "full_name"              => full_name,
+      "tap"                    => tap&.name,
+      "oldname"                => oldnames.first, # deprecated
+      "oldnames"               => oldnames,
+      "aliases"                => aliases.sort,
+      "versioned_formulae"     => versioned_formulae.map(&:name),
+      "desc"                   => desc,
+      "license"                => SPDX.license_expression_to_string(license),
+      "homepage"               => homepage,
+      "versions"               => {
         "stable" => stable&.version&.to_s,
         "head"   => head&.version&.to_s,
         "bottle" => bottle_defined?,
       },
```

Comment:
```suggestion
      "stable_version"               => stable&.version&.to_s,
      "head_version"                 => head&.version&.to_s,
      "bottle_defined"               => bottle_defined?,
```

to kill some nesting

Reply:
`head_version` and `bottle_defined` seems unnecessary to have at all.

IMO we should have a `stable` hash that has url version sha256 together.

Reply:
> `head_version` and `bottle_defined` seems unnecessary to have at all.

I think I said earlier but just to restate: we should not have any keys included here that are not used/needed at all in the `Formulary` API loader.

> IMO we should have a `stable` hash that has url version sha256 together as that aligns with how the DSL works:

I agree (I think?) with @apainintheneck that for the internal API: I think it's nicer to have this be either nicely nested everywhere or entirely flat everywhere rather than half and half. I am assuming that entirely flat is "more efficient" to download and/or read into memory; if this is not the case (or the opposite is the case): let's nest everything.

I also think the public API warrants a (separate, later) debate about human readability vs. programmatic parsing vs. size. 

My $0.02 as a consumer of the Homebrew API many times over across a bunch of languages at this point: JSON nesting always makes my life harder rather than easier for a very, very small "this looks nice".

Reply:
> My $0.02 as a consumer of the Homebrew API many times over across a bunch of languages at this point: JSON nesting always makes my life harder rather than easier for a very, very small "this looks nice".

That's because it was backwards a lot of the time

> I am assuming that entirely flat is "more efficient" to download and/or read into memory

In theory yes, but we're getting into territory where it's so marginal that it might not even be measurable. At some point we probably just need to go with what we've got and try iterative PRs to see what actually makes a difference.

Reply:
> That's because it was backwards a lot of the time

Sorry, to clarify: I'd also extend this to "literally any nesting at any time".

> We're probably at the point where we probably just need to go with what we've got as a baseline and try iterative PRs to see what actually makes a difference. Ultimately, benchmarks is the reason we started this PR in the first place and we've not seen any benchmarks on any of the changes so far yet.

That works for me, just wanted to make sure folks know I may not want to roll this out fully until I've gone a final pass of the JSON like this 👍🏻

Reply:
This is in the current external API which is out of the scope for this PR. If we want to change it, we will have to go through the whole deprecation cycle because it will be a breaking change.

The only change to the output of the `#to_hash` method will be keys changing order which shouldn't matter when parsing JSON. This change is already mentioned in the PR description.

Reply:
> > I am assuming that entirely flat is "more efficient" to download and/or read into memory
> 
> In theory yes, but we're getting into territory where it's so marginal that it might not even be measurable. How this all affects gzip compression done by GitHub Pages is another factor. The frozen string change I made should also in theory reduce the effect of duplicated strings.
> 
> We're probably at the point where we probably just need to go with what we've got as a baseline and try iterative PRs to see what actually makes a difference. Ultimately, benchmarks is the reason we started this PR in the first place and we've not seen any benchmarks on any of the changes so far yet.

100% agree with this. As long as we don't change the old format in this PR and the `brew generate-formula-api` command is not too slow (those are benchmarks I need to add to this PR) I think we're good.

Reply:
> This is in the current external API which is out of the scope for this PR. If we want to change it, we will have to go through the whole deprecation cycle because it will be a breaking change.

@apainintheneck I agree we should not change the existing external API. It is/was not obvious from this diff where the external API is being changed and when the new one is changed. As I mentioned before: I don't think we should be blocked on or relying on formatting from the existing API when creating this one. If it seems like I'm suggesting breaking the existing API: I am not, I am suggesting the change in the new one.

Reply:
The output of `Formula#to_api_hash` reflects the new internal API format, The `Formula#to_hash` output needs to remain backwards compatible since it's used to generate the current API and the output for `brew info --json=`. All changes to `Formula#to_hash` in the diff are related to sharing logic with `Formula#to_api_hash` for things that shouldn't differ too much between them. We could potentially separate them further but that would mean more code duplication.

The suggestion in this thread would change the format of the `Formula#to_hash` method which is off limits here. Let me know if there is anything else that is unclear.

Reply:
Ok, thanks for explaining ❤️.

I think there may be a benefit in separating them more but that doesn't need to block this PR.